### PR TITLE
chore: Standarise fetch in CLI

### PR DESCRIPTION
**Pull request by Esemesek**
Summary:
---------

Standardise  `fetch` calls in CLI.

Fixes https://github.com/react-native-community/cli/issues/423

Test Plan:
----------

- [x] - Make sure that tests pass
- [x] - Make sure that commands using new fetch function work on real world example

### CODE REVIEW

File: packages/cli/src/commands/init/__tests__/templateName.test.js [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
+import {fetch} from '@react-native-community/cli-tools';
 
-jest.mock('../../../tools/fetch', () => ({fetch: jest.fn()}));
+jest.mock('@react-native-community/cli-tools', () => ({fetch: jest.fn()}));
 
 const RN_NPM_PACKAGE = 'react-native';
 const ABS_RN_PATH = '/path/to/react-native';
 
+const mockFetch = (value: any, status: number) => {
+  (fetch: any).mockImplementation(() =>
```
Comment (thymikee) [inline packages/cli/src/commands/init/__tests__/templateName.test.js]: this was mockImplementationOnce

File: packages/cli/src/commands/init/templateName.js [hunk 2]
```diff
@@ -81,7 +81,13 @@ async function tryTemplateShorthand(templateName: string) {
       `https://registry.yarnpkg.com/${reactNativeTemplatePackage}/latest`,
     );
 
-    if (JSON.parse(response).name) {
+    if (response.status < 200 || response.status > 299) {
+      throw new Error(`Failed to load page, status code: ${response.status}`);
+    }
+
+    const body = await response.json();
```
Comment (thymikee) [inline packages/cli/src/commands/init/templateName.js]: can't we do the `response.json` inside fetch function?

File: packages/cli/src/commands/upgrade/__tests__/upgrade.test.js [hunk 3]
```diff
@@ -217,7 +225,7 @@ test('fetches regular patch, adds remote, applies patch, installs deps, removes
   );
 }, 60000);
 test('fetches regular patch, adds remote, applies patch, installs deps, removes remote when updated from nested directory', async () => {
-  (fetch: any).mockImplementation(() => Promise.resolve(samplePatch));
+  mockFetch(samplePatch, 200);
```
Comment (thymikee) [inline packages/cli/src/commands/upgrade/__tests__/upgrade.test.js]: looks like we could use single mock in testing tools

File: packages/cli/src/commands/init/templateName.js (hunk 2 above)
Comment (Esemesek) [reply packages/cli/src/commands/init/templateName.js]: Yes, but sometimes we check for status of the request. Maybe we should add generic handler if the response status code is not 2xx?

File: packages/cli/src/commands/init/templateName.js [hunk 4]
```diff
@@ -81,7 +81,13 @@ async function tryTemplateShorthand(templateName: string) {
       `https://registry.yarnpkg.com/${reactNativeTemplatePackage}/latest`,
     );
 
-    if (JSON.parse(response).name) {
+    if (response.status < 200 || response.status > 299) {
+      throw new Error(`Failed to load page, status code: ${response.status}`);
```
Comment (thymikee) [inline packages/cli/src/commands/init/templateName.js]: It's a package, not page. Add some information about it

File: packages/cli/src/commands/init/templateName.js (hunk 2 above)
Comment (thymikee) [reply packages/cli/src/commands/init/templateName.js]: we can check for status inside fetch and then return something like `{data: response.json(), status}`

File: packages/cli/src/commands/init/templateName.js (hunk 4 above)
Comment (Esemesek) [reply packages/cli/src/commands/init/templateName.js]: This error message is copy pasted from previous error message that was here before. We can add some generic error handling in this case. I started a conversation in one of the previous comment. Closing this one.

File: packages/cli/src/commands/init/templateName.js (hunk 2 above)
Comment (Esemesek) [reply packages/cli/src/commands/init/templateName.js]: That makes sense. On it.

**Comment by thymikee** [comment]
ping

**Comment by Esemesek** [comment]
@thymikee I'll finish this likely today C:

File: packages/cli/src/commands/upgrade/upgrade.js [hunk 5]
```diff
@@ -46,7 +45,14 @@ const getPatch = async (currentVersion, newVersion, config) => {
   logger.info(`Fetching diff between v${currentVersion} and v${newVersion}...`);
 
   try {
-    patch = await fetch(`${rawDiffUrl}/${currentVersion}..${newVersion}.diff`);
+    const {status, data} = await fetch(
+      `${rawDiffUrl}/${currentVersion}..${newVersion}.diff`,
+    );
+
+    if (status < 200 || status > 299) {
+      throw new Error(`Failed to load page, status code: ${status}`);
```
Comment (thymikee) [inline packages/cli/src/commands/upgrade/upgrade.js]: Users of `upgrade` command have no idea what page failed to load, status code doesn't tell much as well. Can you make this error a bit more descriptive, and maybe show the error message as well? Also, use `CLIError`.

**Pull request merged**